tests/infra/ccf.py

```diff
@@ -148,144 +148,98 @@ def start_and_join(self, args):
-                        workspace=args.workspace,
-                        label=args.label,
-                        members_certs="member*_cert.pem",
-                        **forwarded_args,
-                    )
+                    if not recovery:
+                        node.start(
+                            lib_name=args.package,
+                            workspace=args.workspace,
+                            label=args.label,
+                            members_certs="member*_cert.pem",
+                            **forwarded_args,
+                        )
+                    else:
+                        node.recover(
+                            lib_name=args.package,
+                            ledger_file=ledger_file,
+                            sealed_secrets=sealed_secrets,
+                            workspace=args.workspace,
+                            label=args.label,
+                            **forwarded_args,
+                        )
                     node.network_state = NodeNetworkState.joined
                 else:
-                    self.add_node(node, args.package, None, args)
+                    self._add_node(node, args.package, args)
             except Exception:
                 LOG.exception("Failed to start node {}".format(i))
                 raise
         LOG.info("All remotes started")
 
-        if primary is None:
-            primary = self.nodes[0]
-
+        primary, _ = self.find_primary()
         self.wait_for_all_nodes_to_catch_up(primary)
-        LOG.success("All nodes joined network")
-
-        self.add_users(primary, initial_users)
-        LOG.info("Initial set of users added")
-
-        self.open_network(primary)
-        LOG.success("***** Network is now open *****")
-
-        return primary, self.nodes[1:]
+        self.check_for_service(primary, status=ServiceStatus.OPENING)
+        return primary
 
-    def start_in_recovery(self, args, ledger_file, sealed_secrets):
-        hosts = self.hosts or ["localhost"] * number_of_local_nodes()
+    def start_and_join(self, args):
+        # TODO: The node that starts should not necessarily be node 0
+        cmd = ["rm", "-f"] + glob("member*.pem")
+        infra.proc.ccall(*cmd)
 
-        if not args.package:
-            raise ValueError("A package name must be specified.")
+        self.create_members([1, 2, 3])
+        initial_users = [1, 2, 3]
+        self.create_users(initial_users)
 
         if args.app_script:
             infra.proc.ccall("cp", args.app_script, args.build_dir).check_returncode()
         if args.gov_script:
             infra.proc.ccall("cp", args.gov_script, args.build_dir).check_returncode()
         LOG.info("Lua scripts copied")
 
-        LOG.info("Starting nodes on {}".format(hosts))
-
-        forwarded_args = {
-            arg: getattr(args, arg) for arg in infra.ccf.Network.node_args_to_forward
-        }
+        primary = self._start_all_nodes(args)
```

Comment:
Some refactoring here as `start_and_join()` and `start_in_recovery()` had become almost identical.